### PR TITLE
V1

### PR DESCRIPTION
**Pull request by ryansmith3136**
Lpxc started off as a reference implementation for Heroku Add-on providers to use as an example. However, there have been several companies (including Heroku) who have started to use this code in production. Thus, I have been motivated to add the following features to the library in hopes of making it a suitable choice for production systems.
- Tests - There are a couple unit tests and an integration test.
- Multi-tenancy - It is now possible to use 1 instance of **lpxc** to transport logs for many logplex tokens.

I am also planning on merging in the changes from #3 into the V1 release.

cc: @deafbybeheading @technomancy

### CODE REVIEW

File: lpxc.rb [hunk 1]
```diff
@@ -6,40 +6,59 @@
...
+    @hash_lock = Mutex.new
+
+    #Ignored by logplex
+    @structured_data = opts[:structured_data] || "-"
+
+    #Ignored by logplex
+    @msgid = opts[:msgid] || "-"
+
+    #This will show up in the Heroku logs tail command as: app[lpxc]
+    @procid = opts[:procid] || "lpxc"
```
Comment (msakrejda) [inline lpxc.rb:31]: Nice.

File: lpxc.rb [hunk 2]
```diff
@@ -6,40 +6,59 @@
...
+
+    #Determines how long to keep the tcp connection to logplex alive.
+    @conn_timeout = opts[:conn_timeout] || 60
+
+    #Number of factional seconds to batch messages in memory.
+    @flush_interval = opts[:flush_interval] || 0.5
+
+    #In most cases, the value should be: https://east.logplex.io/logs
+    err = "Must set logplex url."
+    @logplex_url = URI(opts[:logplex_url] || ENV["LOGPLEX_URL"] || raise(err))
```
Comment (msakrejda) [inline lpxc.rb]: Also nice.

File: lpxc.rb [hunk 3]
```diff
@@ -6,40 +6,59 @@
...
+      q.enq({t: Time.now.utc, token: tok, msg: msg})
+    end
   end
 
   #This method must be called in order for the messages to be sent to Logplex.
   #This method also spawns a thread that allows the messages to be batched.
   #Messages are flushed from memory every 500ms or when we have 300 messages,
   #whichever comes first.
-  def self.start
+  def start
```
Comment (msakrejda) [inline lpxc.rb]: Is `#start` necessary? Given that there is no matching stop, or any other semantics for controlling the worker threads, can't this just happen implicitly on first `puts`? Is that crazy?

File: lpxc.rb [hunk 4]
```diff
@@ -48,61 +67,75 @@ def self.start
...
-          http.use_ssl = true
+          http.use_ssl = true if @logplex_url.scheme == 'https'
           http.start do |conn|
             loop do
+              #We will block here waiting for a request.
               req = @reqs.deq
               req.add_field('Content-Type', 'application/logplex-1')
               resp = conn.request(req)
-              $stdout.puts("at=request-sent status=#{resp.code}")
+              $stdout.puts("at=req-sent status=#{resp.code}") if ENV['DEBUG']
```
Comment (msakrejda) [inline lpxc.rb]: Nice.

File: readme.md [hunk 5]
```diff
@@ -1,19 +1,17 @@
...
+A simple ruby client to help send log messages to Heroku's logplex.
 
 ## Usage
 
 ```ruby
 require 'lpxc'
-Lpxc.Start
-Lpxc.puts(logplex_token, "hello world")
+lpxc = Lpxc.new
+lpxc.puts("hello world", 't.123')
```
Comment (msakrejda) [inline readme.md:14]: Unless you decide to make `#start` implicit as per the above suggestion, it should be restored here.

**Comment by msakrejda** [comment]
Excellent. Despite the relative rawness, `lpxc` has been working pretty well for us so far. I'm very excited by having a more mature structure to the project.

File: lpxc.rb (hunk 1 above)
Comment (technomancy) [inline lpxc.rb:31]: Yeah, this is better the way in #3 which just takes procid as an arg to `#start`.

File: lpxc.rb [hunk 6]
```diff
@@ -48,61 +67,75 @@ def self.start
...
   end
 
-  #Format the user message into rfc5425 format.
+  #Format the user message into RFC5425 format.
   #This method also prepends the length to the message.
-  def self.fmt(data)
+  def fmt(data)
     pkt = "<190>1 "
-    pkt += "#{data[:ts]} "
+    pkt += "#{data[:t].to_datetime.rfc3339.to_s} "
```
Comment (technomancy) [inline lpxc.rb]: The `to_datetime` method is new in 1.9, so you'll want to use the `strftime` invocation from #3 here instead. That and the literal hash syntax in `puts` were the only things preventing bamboo compatibility IIRC.

**Comment by ryansmith3136** [comment]
@technomancy Thanks for the pointers on 1.8.7 compatibility. I implemented your suggestions and the test are now passing on that version of ruby.

**Comment by ryansmith3136** [comment]
Before we cut the first version of this code, I would like to address the start/stop issue. I am not sure I know the best way to handle this. I am going to play around with a few different approaches and propose a suggestion later today.

**Comment by ryansmith3136** [comment]
Thanks for the feedback guys. I continued to make improvements and cut a release.

**Pull request merged**